java-util/src/main/java/io/druid/java/util/common/RetryUtils.java

```diff
@@ -66,9 +76,12 @@
     while (true) {
       try {
         nTry++;
-        return f.call();
+        return f.perform();
       }
       catch (Throwable e) {
+        if (onEachFailure != null) {
+          onEachFailure.perform();
```

Comment:
If this is "cleanupAfterFailure", could catch exceptions and just log them, but not propagating

Reply:
Probably users should be able to choose throw an exception on error or ignore it. I added a javadoc.